hooks/tk-multi-publish2/maya/publish_files_integ.py

```diff
@@ -0,0 +1,370 @@
+
+CAMERA_NAME = 'CAM'
+
+DEFAULT_CAMERAS = ['persp',
+                   'top',
+                   'front',
+                   'side']
+
+
+class MayaPublishFilesDDIntegValidationPlugin(HookBaseClass):
+    """
+    Inherits from MayaPublishPlugin
+    """
+
+    @property
+    def description(self):
+        """
+        Verbose, multi-line description of what the plugin does. This can
+        contain simple html for formatting.
+        """
+
+        desc = super(MayaPublishFilesDDIntegValidationPlugin, self).description
+
+        return desc + "<br><br>" + """
+        Validation checks before a file is published.
+        """
+
+    def _build_dict(self, seq, key):
+        """
+        Creating a dictionary based on a key.
+
+        :param seq: list of dictionaries
+        :param key: dictionary key from which to create the dictionary
+        :return: dict with information arranged based on that particular key
+        """
+        return dict((d[key], dict(d, index=index)) for (index, d) in enumerate(seq))
+
+    def _framerange_of_sequence(self, item):
+        """
+        Since users have the option to render only a subset of frames,
+        adding validation to check if the full frame range is being published.
+
+        :param item: Item to process
+        :return: True if yes false otherwise
+        """
+        lss_path = item.properties['path']
+        lss_data = frangetools.getSequence(lss_path)
+
+        info_by_path = self._build_dict(lss_data, key="path")
+        missing_frames = info_by_path.get(lss_path)['missing_frames']
+
+        if missing_frames:
+            self.logger.error("Incomplete playblast! All the frames are not the playblast.")
+            return False
+        else:
+            # If there are no missing frames, checking if the start and end frames match with playblast settings.
+            # This is being directly checked with playblast settings in the scene since
+            # _sync_frame_range_with_shotgun() will ensure playblast frame range is synced with shotgun
+            import pymel.core as pm
+            playback_start = pm.playbackOptions(q=True, minTime=True)
+            playback_end = pm.playbackOptions(q=True, maxTime=True)
+            collected_playblast_firstframe = info_by_path.get(lss_path)['frame_range'][0]
+            collected_playblast_lastframe = info_by_path.get(lss_path)['frame_range'][1]
+            if (collected_playblast_firstframe != playback_start) or (collected_playblast_lastframe != playback_end):
+                self.logger.error("Incomplete playblast! All the frames are not in the playblast.")
+                return False
+        return True
+
+
+    def _sync_frame_range_with_shotgun(self, item):
+        """
+        Checks whether frame range is in sync with shotgun.
+
+        :param item: Item to process
+        :return: True if yes false otherwise
+        """
+        context = item.context
+        entity = context.entity
```

Comment:
What are the cases where context could be invalid?

Reply:
You're right...context will always be valid since the context property will traverse up the item hierarchy if not defined locally, however it is quite possible for entity to be invalid, i.e. if this is a Project-level item, and it is also possible for the entity to not have sg_head_in/sg_tail_out attributes in its schema (i.e. its a Sequence or Asset).